cmd/api/acrsdk.go

```diff
@@ -0,0 +1,218 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package api
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"strings"
+
+	acrapi "github.com/AzureCR/acr-cli/acr"
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+)
+
+const (
+	prefixHTTPS = "https://"
+	registryURL = ".azurecr.io"
+)
+
+// BasicAuth returns the username and the passwrod encoded in base 64
+func BasicAuth(username string, password string) string {
+	auth := username + ":" + password
+	return "Basic " + base64.StdEncoding.EncodeToString([]byte(auth))
+}
+
+// LoginURL returns the FQDN for a registry
+func LoginURL(registryName string) string {
+	// TODO: if the registry is in another cloud (i.e. dogfood) a full FQDN for the registry should be specified.
+	if strings.Contains(registryName, ".") {
+		return registryName
+	}
+	return registryName + registryURL
+}
+
+// LoginURLWithPrefix return the hostname of a registry
+func LoginURLWithPrefix(loginURL string) string {
+	urlWithPrefix := loginURL
+	if !strings.HasPrefix(loginURL, prefixHTTPS) {
+		urlWithPrefix = prefixHTTPS + loginURL
+	}
+	return urlWithPrefix
+}
+
+// AcrListTags list the tags of a repository with their attributes
+func AcrListTags(ctx context.Context,
+	loginURL string,
+	auth string,
+	repoName string,
+	orderBy string,
+	last string) (*acrapi.TagAttributeList, error) {
+	hostname := LoginURLWithPrefix(loginURL)
+	client := acrapi.NewWithBaseURI(hostname,
+		repoName,
+		"",
+		"",
+		"",
+		"",
+		auth,
+		orderBy,
+		"100",
+		last,
+		"")
+	tags, err := client.AcrListTags(ctx)
+	if err != nil {
+		return nil, err
+	}
+	var listTagResult acrapi.TagAttributeList
+	switch tags.StatusCode {
+	case http.StatusOK:
+		if err = mapstructure.Decode(tags.Value, &listTagResult); err != nil {
+			return nil, err
+		}
+		return &listTagResult, nil
+
+	case http.StatusUnauthorized, http.StatusNotFound:
+		var apiError acrapi.Error
+		if err = mapstructure.Decode(tags.Value, &apiError); err != nil {
+			return nil, errors.Wrap(err, "unable to decode error")
+		}
+		if apiError.Errors != nil && len(*apiError.Errors) > 0 {
+			return nil, fmt.Errorf("%s %s", *(*apiError.Errors)[0].Code, *(*apiError.Errors)[0].Message)
+		}
+		return nil, errors.New("unable to decode apiError")
```

Comment:
>unable to decode apiError [](start = 26, length = 25)

If the code reach here, does it mean the decode succeeds but Errors are empty?

Reply:
It means that the decoding was unsuccessful and that there was an error parsing Errors